BUG: Fix DatetimeIndex timezone preservation when joining indexes with same timezone but different units

myenugula opened the pull request:
- [x] closes #60080
- [x] [Tests added and passed]
- [x] All [code checks passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit).
- [x] Added [type annotations]
- [x] Added an entry in the latest `doc/source/whatsnew/vX.X.X.rst` file if fixing a bug or adding a new feature.

myenugula committed:
BUG: Fix DatetimeIndex timezone preservation when joining indexes with same timezone but different units

myenugula commented:
Hi @rhshadrach, could you please check this one out?

rhshadrach commented on pandas/tests/indexes/datetimes/test_setops.py:
```diff
@@ -201,6 +201,46 @@ def test_union_same_timezone_different_units(self):
+        assert idx1.tz == idx2.tz
+        assert idx1.dtype != idx2.dtype  # Different units
+
+        # Test union preserves timezone when units differ
+        result = idx1.union(idx2)
+        expected = date_range("2000-01-01", periods=3, tz=tz).as_unit("ns")
+        tm.assert_index_equal(result, expected)
+        assert result.tz == idx1.tz  # Original timezone is preserved
```
Can you either break these up into separate tests or parametrize them.

rhshadrach reviewed (changes requested):
@mroeschke - this change will make it so the unit comes out to be the minimum between the two time inputs. Does that look like the correct behavior to you?

jbrockmendel commented:
I agree with @rhshadrach's comment on splitting/parametrizing the test, otherwise this LGTM

myenugula committed:
TST: Split timezone preservation test into separate tests

Address review comments on PR #60080 by splitting the comprehensive test into
separate focused tests for each set operation (union, intersection, symmetric_difference).

myenugula committed:
Merge branch 'main' into fix-issue-60080

rhshadrach commented on pandas/tests/indexes/datetimes/test_setops.py:
```diff
@@ -201,6 +201,73 @@ def test_union_same_timezone_different_units(self):
+        result = idx1.union(idx2)
+        expected = date_range("2000-01-01", periods=3, tz=tz).as_unit("ns")
+        tm.assert_index_equal(result, expected)
+        assert result.tz == idx1.tz  # Original timezone is preserved
```
Can you remove `assert result.tz == idx1.tz` throughout; this is already checked by `assert_index_equal`.

myenugula replied:
I removed all of them - sorry wasn't aware of that

myenugula committed:
Remove unnecessary assert result.tz == idx1.tz

myenugula commented:
Hi @rhshadrach, I see that you've requested changes. Could you please clarify what exactly needs to be changed? as I've already made the changes you've requested about `assert result.tz == idx1.tz`

rhshadrach commented:
> Hi @rhshadrach, I see that you've requested changes. Could you please clarify what exactly needs to be changed? as I've already made the changes you've requested about `assert result.tz == idx1.tz`

Once changes are requested, the state doesn't update until another review is submitted.

rhshadrach reviewed (approved):
lgtm
